localization/RingBufferFuncs.py

```diff
@@ -0,0 +1,25 @@
+"""Basic ringbuffer functions. Ref from Sébastien Keim @ Oreilly"""
+
+
+class RingBuffer:
+    """ class that implements a not-yet-full buffer """
+    def __init__(self,size_max):
+        self.max = size_max
+        self.data = []
+
+    class __Full:
```

Comment:
A little confused that there is a class defined in RingBuffer Class😂 How is it constructed? And where is this class used? Have not seen code using this nested class.

Reply:
RingBuffer is the main class, and it switches to class Full by "self.class = self.__Full" when the buffer is full.